src/main/flight/autopilot.c

```diff
@@ -91,17 +215,186 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+        vector2_t pidDA;                // D+A
+
+        for (axisEF_e efAxisIdx = LON; efAxisIdx <= LAT; efAxisIdx++) {
+            efPidAxis_t *efAxis = &ap.efAxis[efAxisIdx];
+            // separate PID controllers for longitude (EastWest or EW, X) and latitude (NorthSouth or NS, Y)
+            const float axisDistance = gpsDistance.v[efAxisIdx];
+
+            // ** P **
+            const float pidP = axisDistance * positionPidCoeffs.Kp;
+            pidSum.v[efAxisIdx] += pidP;
+
+            // ** I **
+            // only add to iTerm while in hold phase
+            efAxis->integral += efAxis->isStopping ? 0.0f : axisDistance * gpsDataInterval;
+            const float pidI = efAxis->integral * positionPidCoeffs.Ki;
+            pidSum.v[efAxisIdx] += pidI;
+
+            // ** D ** //
+            // Velocity derived from GPS position works better than module supplied GPS Speed and Heading information
+
+            const float velocity = (axisDistance - efAxis->previousDistance) * gpsDataFreq; // cm/s
+            efAxis->previousDistance = axisDistance;
+            pt1FilterUpdateCutoff(&efAxis->velocityLpf, vaGain);
+            const float velocityFiltered = pt1FilterApply(&efAxis->velocityLpf, velocity);
+            float pidD = velocityFiltered * positionPidCoeffs.Kd;
+
+            // differentiate velocity another time to get acceleration
+            float acceleration = (velocityFiltered - efAxis->previousVelocity) * gpsDataFreq;
+            efAxis->previousVelocity = velocityFiltered;
+            // apply second filter to acceleration (acc is filtered twice)
+            pt1FilterUpdateCutoff(&efAxis->accelerationLpf, vaGain);
+            const float accelerationFiltered = pt1FilterApply(&efAxis->accelerationLpf, acceleration);
+            const float pidA = accelerationFiltered * positionPidCoeffs.Kf;
+
+            if (ap.sticksActive) {
+                // sticks active 'phase', prepare to enter stopping
+                efAxis->isStopping = true;
+                // slowly leak iTerm away
+                efAxis->integral *= iTermLeakGain;
+                efAxis->previousDistance = 0.0f; // avoid D and A spikes
+                // rest is handled after axis loop
+            } else if (efAxis->isStopping) {
+                // 'phase' after sticks are centered, but before craft has stopped in given PID axis
+                pidD *= 1.6f; // aribitrary D boost to stop more quickly when sticks are centered
+                // detect velocity zero crossing (velocityFiltered is delayed by filter)
+                if (velocity * velocityFiltered < 0.0f) {
+                    // when an axis has nearly stopped moving, reset it and end it's start phase
+                    const int8_t llhAxisInv = 1 - efAxisIdx; // because we have Lat first in gpsLocation_t, but efAxisIdx handles lon first.
+                    ap.targetLocation.coords[llhAxisInv] = gpsSol.llh.coords[llhAxisInv]; // forcing P to zero
```

Comment:
But that won't work, because gpsSol is latitude, longitude, height, in that order, and efAxisIdx is longitude, latitude, in that order.  So we can't do that otherwise we will cross latitude into longitude.
Hence have to switch the axes around here.
It's a great shame that gpsSol.llh uses 
``
typedef union gpsLocation_u {
    struct {
        int32_t lat;                // latitude * 1e+7
        int32_t lon;                // longitude * 1e+7
        int32_t altCm;              // altitude in 0.01m
    };
    int32_t coords[3];              // added to provide direct access within loops
} gpsLocation_t;
```
If we want ef consistency we should either do lat first, or lon first, everywhere.
In this function we see the problem very nicely.
I also have it in the gps rescue code coming up.  Maybe there we can resolve this once and for all.

Reply:
setTargetLocationByAxis takes axisEF_e argument and uses it correctly, with no dependence on gpsLocation format.
``` 
void setTargetLocationByAxis(const gpsLocation_t* newTargetLocation, axisEF_e efAxisIdx) {
    if (efAxisIdx == LON) {
            ap.targetLocation.lon = newTargetLocation->lon;
```

As I said multiple times - gpsLocation is value in polar coordinates (with 6378km offset for height). Is is **not** in NEU system. 
Distance in this algorithm is cartesian system value, in EN(U) format. X is almost matching longitude, but not exactly. Mixing these 2 systems is possible because the errors are negligible and does not cause significant problems, at least with current code. But it IMO better to keep the systems separated (using `GPS_distance2d` to convert from first to the second).

Reply:
BTW: poshold on Fiji will be fun .... taking the 40kkm long way to reach target ....

Reply:
@ctzsnooze : ?

Reply:
Sorry, @ledvinap, I don't understand what's wrong with the current code.
It works properly, has been flight tested.
If I implement your proposed alternative, the code does not work unless I use an external function.
The way the code is now, no external function is required, we write the current loop value directly to the appropriate index in the gpsLocation_t struct.
The comment explains that the gpsLocation struct has  [0] as latitude,  [1] as longitude, and [2] for height.
This loop is being run with iteration [0] evaluating longitude, while iteration [1] evaluates latitude.  A value from iteration [0] of the loop needs to be assigned to longitude, ie index [1] of gpsLocation.  The subtraction 1 - current index of the loop converts [0] to [1].
It was your original suggestion to make the order of the for loop as it is now (longitude first, then latitude).  That's why I originally went latitude first, then longitude, in earlier versions of this code.

Reply:
Yes, the code works, but:
- it does work for 2d case only, but pretends that is is generic
- it does depend on llh layout and its internals
- it pretends that llh is homogenous 3 element vector. But llh is angle/angle/distance - it is very reasonable to redefine it as `int32_t`, `int32_t`, `float` [in meters]

> If I implement your proposed alternative, the code does not work unless I use an external function.

`setTargetLocationByAxis` is part of this file. It will be inlined here, so overhead is minimal. It is quite easy to read and understand, it is idempotent of LLH ordering or axisEF_e LAT/LON definition.

Reply:
@ledvinap OK... I've implemented the `setTargetLocationByAxis` method. I agree that it is easier to read and understand.  
Ideally we should have gps data enum order match vector maths order for latitude and longitude.  That would simplify matters. Perhaps that can be addressed later.